query-engine/connectors/mongodb-query-connector/src/join.rs

```diff
@@ -79,48 +79,60 @@ impl JoinStage {
         let right_model = from_field.related_model();
         let right_coll_name = right_model.db_name();
 
-        let mut left_scalars = from_field.left_scalars();
-        let mut right_scalars = from_field.right_scalars();
-
         // +1 for the required match stage, the rest is from the joins.
         let mut pipeline = Vec::with_capacity(1 + nested_stages.len());
 
-        // todo: multi-field joins
-        // Field on the right hand collection of the join.
-        let right_field = right_scalars.pop().unwrap();
-        let right_name = right_field.db_name().to_string();
+        // First we start with the right side of the equation
+        let right_scalars = from_field.right_scalars();
+
+        // What $expr operators we will need to express this lookup? (depends on right fields)
+        let ops: Vec<Document> = right_scalars
+            .iter()
+            .enumerate()
+            .map(|(idx, right_field)| {
+                let right_ref = format!("${}", right_field.db_name());
+                let left_var = format!("$$left_{}", idx);
+
+                match relation.is_many_to_many() {
+                    true if right_field.is_list => doc! { "$in": [left_var, right_ref] },
+                    true => doc! { "$in": [right_ref, left_var] },
+                    _ => doc! { "$eq": [right_ref, left_var] },
+                }
+            })
+            .collect();
 
-        // Field on the left hand collection of the join.
-        let left_field = left_scalars.pop().unwrap();
-        let left_name = left_field.db_name();
+        // If the relationship is many to many we need to push an $addFields to the pipeline
+        if relation.is_many_to_many() {
+            // addFields is the list of fields and conditions
+            let mut add_fields = Document::new();
 
-        let right_ref = format!("${}", right_name);
-        let op = if relation.is_many_to_many() {
-            // For m-n join stages: Add an `$addFields` stage that adds an empty array if not present (required to make joins work).
-            pipeline.push(doc! {
-                "$addFields": {
-                    right_name: {
-                        "$cond": {
+            // Go through every right field to place in the $addFields operator
```

Comment:
The comment that explains why this `$addFields` operator is needed is gone. `addFields is the list of fields and conditions` is fairly vague IMO. What fields and conditions, and for what?

The old comment `For m-n join stages: Add an `$addFields` stage that adds an empty array if not present (required to make joins work).` explains what this is for better.